### DISCUSSION TITLE
Create a mapview::mapview() method for sfnetwork objects

### DISCUSSION THREAD

**Discussion by agila5** [luukvdmeer/sfnetworks#126] [Ideas]
My favourite approach for interactively exploring spatial data is [`mapview`](https://r-spatial.github.io/mapview/) (and this probably the first R code that @Robinlovelace taught to me while I was in Leeds 🤣). Unfortunately, at the moment there is no way for directly plotting an `sfnetwork` object with mapview. 

``` r
# packages
library(sf)
#> Linking to GEOS 3.8.0, GDAL 3.0.4, PROJ 6.3.1
library(tidygraph)
#> 
#> Attaching package: 'tidygraph'
#> The following object is masked from 'package:stats':
#> 
#>     filter
library(sfnetworks)
library(mapview)

# sfn data
sfn_roxel <- as_sfnetwork(roxel)

# plot
mapview(sfn_roxel)
#> Error in (function (classes, fdef, mtable) : unable to find an inherited method for function 'mapView' for signature '"sfnetwork"'
```
We can always escape the spatial dimension, for example: 
``` r
mapview(sfn_roxel %>% activate("nodes") %>% st_as_sf(), col.regions = "darkred", legend = NULL) + 
  mapview(sfn_roxel %>% activate("edges") %>% st_as_sf(), color = "black", lwd = 2, legend = NULL)
#> Warning in cbind(`Feature ID` = fid, mat): number of rows of result is not a
#> multiple of vector length (arg 1)
```

![](https://i.imgur.com/zUwOm26.png)

<sup>Created on 2021-02-15 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

but, if you agree, I think that the best solution is to add a new method (here or in mapview package). It's not really that important, but I could work on this addition in the next weeks.

**Comment by loreabad6** [comment]
Nice idea! I think we could do it the same way as the `autoplot.sfnetwork()` method (with a dynamic registration of the S3 method), so that we can put mapview on Suggests :)

**Reply by agila5** [reply]
Thanks! Actually, I think that the `mapview` code is written using S4. Do you know if we can use the same approach?

**Reply by loreabad6** [reply]
🤔 mmm I don't think so, as far as I checked on the `vctrs` package, they don't mention it... let me do some more checks though!

**Comment by agila5** [comment]
A rough implementation with a toy example: 

``` r
library(sfnetworks)
library(mapview)

# Setup
setOldClass("sfnetwork")
setMethod(
  f = "mapView",
  signature = "sfnetwork", 
  definition = function(x, draw_lines = TRUE, ...) {
    nodes <- sf::st_as_sf(x, "nodes")
    m1 <- mapview(nodes, ...)
    if (draw_lines) {
      x <- sfnetworks:::explicitize_edges(x)
      edges <- sf::st_as_sf(x, "edges")
      m2 <- mapview(edges, ...)
      m1 + m2
    } else {
      m1
    }
  }
)

# Toy example
sfn <- as_sfnetwork(roxel)
mapview(sfn, legend = NULL)
```

![image](https://github.com/luukvdmeer/sfnetworks/assets/22221146/2f9bd69c-b95c-469a-9d15-9c885e4bb7c1)

<sup>Created on 2023-07-31 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

I think we can play with it a little bit and then propose it to mapview as a PR (might be easier than adding it to `sfnetworks` but I'm not sure about that).

**Comment by Robinlovelace** [comment]
Could be a good one for `tmap4`? See info on that here: https://github.com/r-tmap/state-of-tmap-2023

**Reply by agila5** [reply]
I would be happy to work on `tm_sfnetwork` (or similar) as soon as tmap v4 is ready and there is a guide on how to write extensions!

**Reply by Robinlovelace** [reply]
cc @mtennekes although you probably have enough on your plate! Hopefully useful to know there are other potential `tm_...()` functions!

**Reply by mtennekes** [reply]
Interesting one. Actually, this is another type of extension (other than the four mentioned https://github.com/r-tmap/state-of-tmap-2023/blob/main/tmap_workshop.md#extensions), namely the support of another spatial data object class.

When methods for sfnetworks have been implemented, standard layers like `tm_symbols` and `tm_lines` should work. E.g.

```r
tm_shape(sfn) +
tm_ lines()
```

Additionally, we can think of `tm_network()` as a new layer which draws nodes and edges.

**Reply by agila5** [reply]
Hi @mtennekes and thank you for your reply. Just one question: what do you mean by "When methods for sfnetworks have been implemented"?

**Reply by mtennekes** [reply]
Good question. Hard to give a short answer. The long one is here: https://github.com/r-tmap/tmap/issues/757 The rudimentary implementation is here https://github.com/r-tmap/tmap/blob/v4/R/tmapShape_sfnetworks.R There are some questions in the comments which you can answer better. Please take a look. To clarify: this implementation will probably not be part of tmap, but we can write an extension package for it.

**Reply by agila5** [reply]
Wow @mtennekes, that looks amazing, thank you very much 🤩 I'll try to answer some of your questions: 

>  are there direct methods to retrieve node and edge variable names?

A possible wrapper (which is more or less the same as what you wrote): 

``` r
# packages
library(sfnetworks)
library(tidygraph)
#> 
#> Attaching package: 'tidygraph'
#> The following object is masked from 'package:stats':
#> 
#>     filter

# toy data
sfn <- as_sfnetwork(roxel)

# define functions
nodes_get_names <- function(x) {
  sf_column <- with_graph(x, attr(.N(), "sf_column"))
  all_names <- igraph::vertex_attr_names(x)
  setdiff(all_names, sf_column)
}
edges_get_names <- function(x) {
  all_names <- igraph:::edge_attr_names(x)
  if (!sfnetworks:::has_explicit_edges(x)) {
    return(all_names)
  }
  sf_column <- with_graph(x, attr(.E(), "sf_column"))
  setdiff(all_names, sf_column)
}
get_names <- function(x, active = NULL) {
  if (is.null(active)) {
    active = attr(x, "active")
  }
  switch(
    active, 
    nodes = nodes_get_names(x), 
    edges = edges_get_names(x),
    sfnetworks:::raise_unknown_input(active)
  )
}

# examples
get_names(sfn, "nodes")
#> character(0)
get_names(sfn, "edges")
#> [1] "name" "type"
```

<sup>Created on 2023-08-12 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

> sfnetwork: are there direct methods to retrieve factor levels?

Not that I'm aware of, but we could also write a wrapper in a similar spirit as the previous example.

**Reply by mtennekes** [reply]
Thx! A new package is born :-) Please check out https://github.com/r-tmap/tmap.sfnetworks Added you as authors, @agila5 and @Robinlovelace

**Reply by Robinlovelace** [reply]
Wow impressive! Andrea best equipped to contribute directly to that I guess but happy to take a peak soon...

**Reply by agila5** [reply]
Hi @mtennekes and thank you very much for your example, it's a great addition! I would be happy to work on that extension after the release of tmap v4 :)

**Reply by mtennekes** [reply]
Thx @agila5 the main reason to get this working at this stage (and not after tmap4 has been released), is that is serves as a POC for tmap extendability (both regarding spatial object class and new layer functions). I'll continue working on it when needed, that is, from a tmap4 point of view. When it is released, then I'll give the floor to you to work on it. Would you also be happy to be the official maintainer in the future when it is released?

**Reply by agila5** [reply]
> Would you also be happy to be the official maintainer in the future when it is released?

Sure thing!
